### PR TITLE
App attestation for code exchange + refresh flow

### PR DESCRIPTION
**Pull request by bbirman**
Created new branch off of dev

This PR includes:
- Attestation structure for key generation & challenge generation
- Flag to enable / disable
- Using attestation on code exchange and refresh

Doesn't include:
- Enforcing web server if attestation is enabled (upcoming PR)
- Feature flag (upcoming PR)
- Native login
- Error handling / retryable errors

### CODE REVIEW

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthSessionRefresher.m [hunk 1]
```diff
@@ -77,13 +78,22 @@ - (void)refreshSessionWithCompletion:(void (^)(SFOAuthCredentials *))completionB
         return;
     }
     
+    __weak typeof(self) weakSelf = self;
+    [SFSDKAppAttestation attestationIfEnabledFor:self.credentials.domain consumerKey:self.credentials.clientId completionHandler:^(NSString * _Nullable attestation) {
+        __strong typeof(weakSelf) strongSelf = weakSelf;
+        [strongSelf executeRefreshWithAttestation:attestation];
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthSessionRefresher.m]: If `weakSelf` has been deallocated by the time the attestation callback fires, `strongSelf` is nil and `[strongSelf executeRefreshWithAttestation:]` is a silent no-op — neither `completionBlock` nor `errorBlock` is ever called. Before this change, `refreshSessionWithCompletion:` was synchronous, so owners may not hold a strong reference across the now-async gap. Callers waiting for a callback will hang indefinitely. Consider explicitly calling the error block if `strongSelf` is nil, or retaining `self` for the duration of the async operation.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m [hunk 2]
```diff
@@ -618,6 +618,12 @@ - (void)updateCredentials:(NSDictionary *) params {
 }
 
 - (void)beginTokenEndpointFlow {
+    [SFSDKAppAttestation attestationIfEnabledFor:self.credentials.domain consumerKey:self.credentials.clientId completionHandler:^(NSString * _Nullable attestation) {
+        [self executeTokenEndpointFlowWithAttestation:attestation];
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m]: This block captures `self` strongly, unlike every other block in this file which uses the `__weak`/`__strong` pattern. During the attestation network round-trip (challenge request), the coordinator is kept alive by the block even if it should have been released (e.g., login cancelled). Should use `__weak typeof(self) weakSelf = self` consistent with the rest of the file.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthSessionRefresher.m (hunk 1 above)
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthSessionRefresher.m]: If `weakSelf` has been deallocated by the time the attestation callback fires, `strongSelf` is nil and `[strongSelf executeRefreshWithAttestation:]` is a silent no-op — neither `completionBlock` nor `errorBlock` is ever called. Before this change, `refreshSessionWithCompletion:` was synchronous, so owners may not hold a strong reference across the now-async gap. Callers waiting for a callback will hang indefinitely. Consider explicitly calling the error block if `strongSelf` is nil, or retaining `self` for the duration of the async operation.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift [hunk 3]
```diff
@@ -0,0 +1,187 @@
...
+        let attestationId: String
+        let keyId: String
+    }
+    
+    static let appAttestKeyName = "com.salesforce.mobilesdk.attestation"
+    
+    /// Returns an attestation string if attestation is enabled and the domain is a My Domain,
+    /// otherwise returns nil. Callers do not need to check gating conditions themselves.
+    @objc
+    public static func attestationIfEnabled(for domain: String, consumerKey: String) async -> String? {
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: The generated ObjC completion handler for this `async` method (`attestationIfEnabledFor:consumerKey:completionHandler:`) fires on Swift's cooperative thread pool, not the main thread. Both `SFOAuthCoordinator` and `SFOAuthSessionRefresher` receive the callback on that background thread and immediately access `self.responseData`, `self.credentials`, `self.approvalCode`, etc. — properties that were only ever touched synchronously before this change. If those callers assume main-thread access (typical for OAuth coordinator flows), this is a data race. Consider dispatching to the expected thread inside the completion handler before the downstream methods are invoked.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+        let challenge = try await requestChallengeFromSalesforce(domain: domain, consumerKey: consumerKey, attestationId: attestationId)
+        let attestation = try await generateAttestation(keyId: keyId, challenge: challenge)
+        try await registerKeyWithSalesforce(keyId: keyId, consumerKey: consumerKey, attestationId: attestationId, attestationObject: attestation, domain: domain)
+
+        // Only persist after successful registration — if any step above fails,
+        // the next attempt will start fresh with a new key.
+        let keychainItem = AttestationKeychainItem(attestationId: attestationId, keyId: keyId)
+        let keychainItemData = try JSONEncoder().encode(keychainItem)
+        let keychainResult = KeychainHelper.write(service: appAttestKeyName, data: keychainItemData, account: nil)
+        if !keychainResult.success {
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift:164]: When the keychain write fails, the error is logged but execution falls through and `keyId` is still returned as if everything succeeded. The key is registered with the server but not persisted locally. On the next launch, `existingKeyId()` returns nil, a new key is generated and registered again, and the previous server-side key becomes orphaned. This will cause a redundant registration on every launch if the keychain is systematically unavailable. Should a keychain write failure be propagated as a thrown error rather than silently continuing?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+            if let error = keychainResult.error {
+                message += ": \(error.localizedDescription)"
+            }
+            SFSDKCoreLogger.log(AppAttestation.self, level: .error, message: message)
+        }
+
+        return keyId
+    }
+
+    private static func performAppAttestOperation<T>(_ operation: () async throws -> T) async throws -> T {
```
Comment (wmathurin) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift:175]: `performAppAttestOperation` only intercepts `DCError` — non-`DCError` throws from `DCAppAttestService` (network errors, cancellation, etc.) pass through here without any logging. The top-level `attestationIfEnabled` catches them and logs `error.localizedDescription`, but context about which specific operation failed is lost. The PR description notes error handling is still a work in progress, so flagging for tracking — consider either logging all errors here, or removing the wrapper in favour of inline logging at each call site.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthSessionRefresher.m (hunk 1 above)
Comment (bbirman) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthSessionRefresher.m]: updated!

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m (hunk 2 above)
Comment (bbirman) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/SFOAuthCoordinator.m]: updated!

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 4 above)
Comment (bbirman) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift:164]: I was originally thinking that regenerating if there was a keychain issue would be less disruptive than blocking login/refresh, but good either way! Should I change it?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 5 above)
Comment (bbirman) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift:175]: added a line for logging other errors

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 3 above)
Comment (bbirman) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: I'm not sure I understand this one, it doesn't look like the token endpoint flow ran on main before, what assumes main thread access?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift [hunk 6]
```diff
@@ -0,0 +1,190 @@
...
+    /// otherwise returns nil. Callers do not need to check gating conditions themselves.
+    @objc
+    public static func attestationIfEnabled(for domain: String, consumerKey: String) async -> String? {
+        let isLoginPool = domain == kSFOAuthProductionLoginURL
+            || domain == kSFOAuthSandboxLoginURL
+            || domain == kSFOAuthWelcomeLoginURL
+
+        guard UserAccountManager.shared.appAttestationEnabled,
+              !isLoginPool,
+              !consumerKey.isEmpty else {
```
Comment (sfdctaka) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: https://developer.apple.com/documentation/devicecheck/dcappattestservice/issupported

On any device where App Attest is unavailable (simulator, older devices, some enterprise MDM configs), `DCAppAttestService.shared.generateKey()` throws `DCError.featureUnsupported` on every single token-endpoint call — you eat an unnecessary network round-trip (challenge request) and two log lines per login/refresh. Guarding on `DCAppAttestService.shared.isSupported` before touching the network would avoid the noise and the wasted request. 
```
  guard UserAccountManager.shared.appAttestationEnabled,
        !isLoginPool,
        !consumerKey.isEmpty,
        DCAppAttestService.shared.isSupported else {
      return nil
  }
```

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 3 above)
Comment (wmathurin) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: Fair point — I was hedging with 'if those callers assume main-thread access.' The real concern is whether callers treat `beginTokenEndpointFlow`/`refreshSessionWithCompletion:` as serialized on a specific thread, because the continuation now lands on Swift's cooperative pool rather than the calling thread. If the ObjC callers already handle these flows on an arbitrary background thread (which sounds like your understanding), there's no issue. Worth a quick check: is there any `dispatch_async(dispatch_get_main_queue(), ...)` wrapping these call sites, or does any caller assume subsequent delegate callbacks arrive on the same thread it called from?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 4 above)
Comment (wmathurin) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift:164]: Your reasoning is sound — failing silently and re-registering on the next launch is less disruptive than blocking auth for a transient keychain failure. No need to change the behavior. The only downside is orphaned server-side keys accumulating if the keychain is persistently broken, but that's an edge case. I'd suggest adding a brief comment at the keychain write failure branch (e.g. `// Intentional: prefer re-registration on next launch over blocking auth`) so future readers understand the choice rather than treating it as an oversight.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 6 above)
Comment (bbirman) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: Good idea! Definitely for the simulator. For older devices / MDM configs, wondering if it should still go through so there's logging around around the failure (especially if the ECA has strict enforcement enabled). What do you guys think?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift [hunk 7]
```diff
@@ -0,0 +1,190 @@
...
+        let attestationId: String
+        let keyId: String
+    }
+    
+    static let appAttestKeyName = "com.salesforce.mobilesdk.attestation"
+    
+    /// Returns an attestation string if attestation is enabled and the domain is a My Domain,
+    /// otherwise returns nil. Callers do not need to check gating conditions themselves.
+    @objc
+    public static func attestationIfEnabled(for domain: String, consumerKey: String) async -> String? {
```
Comment (sfdctaka) [inline libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: Objective-C: `SFOAuthCredentials.domain` is declared as `NSString * _Nullable`— it's legal for it to be
nil.

Swift: `AppAttestation.attestationIfEnabled(for domain: String, ...)` declares domain as a non-optional String — it must have a value.

When ObjC calls into Swift across this bridge, the compiler inserts an implicit "force-unwrap", basically `domain!`. If the ObjC side ever hands in nil, that force-unwrap crashes the app with something like:

`Fatal error: Unexpectedly found nil while implicitly unwrapping an Optional value`

Where this call happens:

Both ObjC call sites do this:
```
// SFOAuthCoordinator.m:622
[SFSDKAppAttestation attestationIfEnabledFor:self.credentials.domain
                                 consumerKey:self.credentials.clientId
                           completionHandler:...];
```

```
// SFOAuthSessionRefresher.m:81  (same shape)
[SFSDKAppAttestation attestationIfEnabledFor:self.credentials.domain consumerKey:self.credentials.clientId completionHandler:^(NSString * _Nullable attestation) //                       If it's nil → crash.
```

Why it's "unlikely but sharp"

In `SFOAuthCredentials.m`:187, domain gets defaulted to "login.salesforce.com" in the initializer. So
in the normal login flow it will never be nil. But:

- A caller could construct/mutate an `SFOAuthCredentials `with `domain = nil` (Swift setter of domain: 
String?).
- A serialized/decoded credentials object with a missing key could produce nil.
- Test doubles or edge-case identity data could set it to nil.

In any of those cases, an otherwise-recoverable auth failure becomes a hard crash inside attestation
gating.

---
Quick fix:
`libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift:53`

Before:
```
@objc
public static func attestationIfEnabled(for domain: String, consumerKey: String) async -> String? {
    let isLoginPool = domain == kSFOAuthProductionLoginURL
        || domain == kSFOAuthSandboxLoginURL
        || domain == kSFOAuthWelcomeLoginURL
        
    guard UserAccountManager.shared.appAttestationEnabled,
          !isLoginPool,
          !consumerKey.isEmpty else {
        return nil
    }
    ...
```

After:
```
@objc
public static func attestationIfEnabled(for domain: String?, consumerKey: String) async -> String? {
    guard let domain, !domain.isEmpty else {
        return nil
    }   
    
    let isLoginPool = domain == kSFOAuthProductionLoginURL
        || domain == kSFOAuthSandboxLoginURL
        || domain == kSFOAuthWelcomeLoginURL
        
    guard UserAccountManager.shared.appAttestationEnabled,
          !isLoginPool,
          !consumerKey.isEmpty else {
        return nil
    }
    ...
```
Two tiny changes:
1. for domain: String → for domain: String?
2. Add guard let domain, !domain.isEmpty else { return nil } at the top.

Now nil-domain simply returns nil (attestation skipped) instead of crashing. The ObjC signature stays
the same, callers don't change.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 6 above)
Comment (sfdctaka) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: Good point. I'm going to defer that to @wmathurin . I'm good either way 👍

Comment (wmathurin) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: I'd split the guard in two:

1. **Simulator**: always guard on `isSupported` — no attestation will ever succeed there and the noise of network calls + log lines on every token refresh is not useful.
2. **Older devices / MDM enforcement**: agree with bbirman — if the admin has enabled strict ECA enforcement, we *want* the failure logged. For those, let it fall through to `performAppAttestOperation` so the `DCError.featureUnsupported` surfaces in the error log with context.

A simple way to handle both: check `isSupported` only in simulator builds (via `#if targetEnvironment(simulator)`), or alternatively check `isSupported && !isSimulator` — but actually the cleanest is to just always run through the normal path and let the logging in `performAppAttestOperation` do its job for real devices. The simulator case is the only one where we'd want a silent early return.

Or simplest of all: add `DCAppAttestService.shared.isSupported` to the existing guard so all unsupported environments are gated out silently, and document in the `appAttestationEnabled` property comment that apps should only enable it on builds with the App Attest entitlement. That makes the developer's intent explicit, and MDM/ECA enforcement scenarios should be handled by the server rejecting the request (not the client pre-screening). What do you think?

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 7 above)
Comment (wmathurin) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: Good catch. sfdctaka's fix is exactly right — change `domain: String` to `domain: String?` and add a `guard let domain, !domain.isEmpty` at the top. The normal login flow won't hit this (domain defaults to `login.salesforce.com` in the credentials initializer), but test doubles, decoded credentials with a missing key, or direct construction with `domain = nil` would otherwise crash instead of skipping attestation gracefully. Please apply the fix.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 6 above)
Comment (wmathurin) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: Clarifying one thing on the ECA concern: ECA (External Client Application) enforcement is configured server-side on the connected app. So if a device has `isSupported = false` and skips attestation, the server will reject the token request with a clear OAuth error — that surfaces through the normal auth error path with a meaningful message. The client-side `DCError.featureUnsupported` log isn't needed for diagnosing ECA misconfiguration; the server response handles that.

So I'd go with adding `DCAppAttestService.shared.isSupported` to the existing guard. The only thing the client-side check buys is skipping the wasted challenge network call on devices that can never produce a valid attestation (simulator, iOS < 14). No diagnostic value is lost.

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 3 above)
Comment (bbirman) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: 👍 double checked, neither

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 6 above)
Comment (bbirman) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: updated!

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 7 above)
Comment (bbirman) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift]: updated!

File: libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift (hunk 4 above)
Comment (bbirman) [reply libs/SalesforceSDKCore/SalesforceSDKCore/Classes/OAuth/AppAttestation.swift:164]: updated!

**Review by sfdctaka — APPROVED** [review]
LGTM!

**Review by wmathurin — APPROVED** [review]
